중고차 가격 예측

model(객체형)
year(정수형)
price(정수형)
transmission(객체형)
mileage(정수형)
fuelType(객체형)
tax(정수형)
mpg(상수형)
engineSize(상수형) 등 총 8개의 변수로 구성되어 있으며, price를 종속변수로 나머지 변수를 독립변수로 설정하여 분석을 진행함.

In [ ]:
import pandas as pd

train = pd.read_csv('/content/drive/MyDrive/데이터 셋/P220504-01.csv')

print(train.shape)
print(train.head())
print(train.info())
print(train.describe())
print(train.isnull().sum())

(6899, 9)
       model  year  price transmission  mileage fuelType  tax   mpg  \
0   5 Series  2014  11200    Automatic    67068   Diesel  125  57.6   
1   6 Series  2018  27000    Automatic    14827   Petrol  145  42.8   
2   5 Series  2016  16000    Automatic    62794   Diesel  160  51.4   
3   1 Series  2017  12750    Automatic    26676   Diesel  145  72.4   
4   7 Series  2014  14500    Automatic    39554   Diesel  160  50.4   

   engineSize  
0         2.0  
1         2.0  
2         3.0  
3         1.5  
4         3.0  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6899 entries, 0 to 6898
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         6899 non-null   object 
 1   year          6899 non-null   int64  
 2   price         6899 non-null   int64  
 3   transmission  6899 non-null   object 
 4   mileage       6899 non-null   int64  
 5   fuelType      6899 non-null   object 
 6   tax         

In [ ]:
print(train['year'].value_counts())
print(train['transmission'].value_counts())
print(train['fuelType'].value_counts())

year
2019    3009
2017    1029
2016     875
2020     622
2018     584
2015     410
2014     229
2013     141
Name: count, dtype: int64
transmission
Semi-Auto    4353
Automatic    1295
Manual       1251
Name: count, dtype: int64
fuelType
Diesel    4323
Petrol    2389
Hybrid     180
Other        7
Name: count, dtype: int64


In [ ]:
test = pd.read_csv('/content/drive/MyDrive/데이터 셋/P220504-02.csv')
print(test.shape)
print(test.head())
print(test.info())
print(test.describe())
print(test.isnull().sum())

(3882, 8)
       model  year transmission  mileage fuelType  tax   mpg  engineSize
0   1 Series  2018    Semi-Auto    14200   Petrol  145  36.7         1.5
1   3 Series  2019    Semi-Auto     2150   Petrol  145  42.2         2.0
2   4 Series  2019    Semi-Auto     6650   Petrol  145  48.7         2.0
3   1 Series  2019    Semi-Auto     2604   Diesel  145  70.6         2.0
4         X3  2016    Semi-Auto    35700   Diesel  145  54.3         2.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3882 entries, 0 to 3881
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   model         3882 non-null   object 
 1   year          3882 non-null   int64  
 2   transmission  3882 non-null   object 
 3   mileage       3882 non-null   int64  
 4   fuelType      3882 non-null   object 
 5   tax           3882 non-null   int64  
 6   mpg           3882 non-null   float64
 7   engineSize    3882 non-null   float64
dtypes: float64(

In [ ]:
train[['year','price']].corr(method='spearman')

year     price
year   1.000000  0.731026
price  0.731026  1.000000

In [ ]:
print(test['year'].value_counts())
print(test['transmission'].value_counts())
print(test['fuelType'].value_counts())

year
2016    1007
2017     692
2015     512
2019     476
2014     272
2018     264
2013     216
2012     119
2020     111
2011      51
2010      41
2009      30
2008      23
2007      16
2006      14
2004      12
2002       6
2005       6
1999       4
2001       3
2003       2
2000       2
1996       1
1997       1
1998       1
Name: count, dtype: int64
transmission
Automatic    2293
Manual       1276
Semi-Auto     313
Name: count, dtype: int64
fuelType
Diesel      2704
Petrol      1028
Hybrid       118
Other         29
Electric       3
Name: count, dtype: int64


In [ ]:
test['fuelType'] = test['fuelType'].replace('Electric', 'Other')

print(test.fuelType.value_counts())

fuelType
Diesel    2704
Petrol    1028
Hybrid     118
Other       32
Name: count, dtype: int64


In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler

mms = MinMaxScaler()
number = ['mileage', 'tax', 'mpg', 'engineSize']
train[number] = mms.fit_transform(train[number])
test[number] = mms.transform(test[number])

ohe = OneHotEncoder(sparse=False)
ohe_train = ohe.fit_transform(train[['transmission','fuelType','year','model']])
ohe_df_train = pd.DataFrame(ohe_train, columns=ohe.get_feature_names_out())
train_df = pd.concat([ohe_df_train, train.drop(['transmission','fuelType','year','model'], axis=1)],axis=1)

ohe_test = ohe.fit_transform(test[['transmission','fuelType', 'year','model']])
ohe_df_test = pd.DataFrame(ohe_test, columns=ohe.get_feature_names_out())
test_df = pd.concat([ohe_df_test, test.drop(['transmission','fuelType','year','model'], axis=1)],axis=1)

print(train_df)
print(test_df)

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


      transmission_Automatic  transmission_Manual  transmission_Semi-Auto  \
0                        1.0                  0.0                     0.0   
1                        1.0                  0.0                     0.0   
2                        1.0                  0.0                     0.0   
3                        1.0                  0.0                     0.0   
4                        1.0                  0.0                     0.0   
...                      ...                  ...                     ...   
6894                     0.0                  0.0                     1.0   
6895                     0.0                  1.0                     0.0   
6896                     0.0                  0.0                     1.0   
6897                     0.0                  0.0                     1.0   
6898                     0.0                  0.0                     1.0   

      fuelType_Diesel  fuelType_Hybrid  fuelType_Other  fuelType_Petrol  \


In [ ]:
features = list(set(train_df.columns.intersection(test_df.columns)))

train = train_df[features]
test = test_df[features]

y = train_df['price']

In [ ]:
print(train.shape)
print(y.shape)

(6899, 42)
(6899,)


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import numpy as np
from sklearn.model_selection import GridSearchCV

x_train, x_valid, y_train, y_valid = train_test_split(train, y,test_size=0.2,random_state=1234)

model = RandomForestRegressor(n_estimators=100, max_depth=20)
model.fit(x_train, y_train)

RandomForestRegressor(max_depth=20)

In [ ]:
params = {'n_estimators':[10, 20, 30, 40 ,50], 'max_depth':[10,20,30,40,50]}

model = RandomForestRegressor()
gscv = GridSearchCV(estimator= model, param_grid = params)
gscv.fit(x_train, y_train)
print(gscv.best_params_)

{'max_depth': 30, 'n_estimators': 50}


In [ ]:
model = RandomForestRegressor(n_estimators=50, max_depth=30)
model.fit(x_train, y_train)
print('결정계수 값:', model.score(x_train, y_train))

pred = model.predict(x_valid)
print(pred)

print('MSE 값:', mean_squared_error(y_valid, pred, squared=False))
rmse = np.sqrt(mean_squared_error(y_valid, pred, squared=False))
print('RMSE 값:', rmse)

결정계수 값: 0.9871391815424149
[20963.64 23430.1  23234.98 ... 20982.84 39701.18 25647.96]
MSE 값: 2787.4338959011593
RMSE 값: 52.79615417718566


모델은 약 0.99의 결정계수 값을 가지며, 독립변수들이 종속변수(price)에 변동을 약99% 설명하고 있다고 볼 수 있다.

mse(평균제곱오차)값은 2787, rmse(평균 제곱근 오차)값은 52으로 비교적 정확한 모델을 구축하였다고 생각됩니다.


In [ ]:
features = list(set(train_df.columns.intersection(test_df.columns)))

test = test_df[features]

pred = model.predict(test)

result = pd.DataFrame({'pred':pred})
print(result)

구축된 모델에 테스트 데이터를 적용하여 값이 정해지지 않은 중고차 price(종속변수)값을 예측해 진행하였음